### PR TITLE
User lib state solution

### PR DESCRIPTION
**Pull request by dunqan**

### CODE REVIEW

File: feature-libs/user/account/core/connectors/user-account.connector.ts [hunk 1]
```diff
@@ -0,0 +1,15 @@
+import { Injectable } from '@angular/core';
+import { Observable } from 'rxjs';
+import { User } from '@spartacus/user/account/root';
+import { UserAccountAdapter } from './user-account.adapter';
+
+@Injectable({
+  providedIn: 'any',
```
Comment (znikola) [inline feature-libs/user/account/core/connectors/user-account.connector.ts]: A reminder to remove this

File: feature-libs/user/package.json [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+    "@angular/router": "^10.1.0",
+    "@spartacus/core": "3.2.0-next.2",
+    "@spartacus/storefront": "3.2.0-next.2",
+    "@spartacus/styles": "3.2.0-next.2",
+    "rxjs": "^6.6.0"
+  },
+  "publishConfig": {
+    "access": "public"
+  },
+  "schematics": "./schematics/collection.json"
```
Comment (znikola) [inline feature-libs/user/package.json]: If we won't add schematics for the user lib, let's remove this line

File: projects/core/src/util/command-query/command.service.ts [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+          ),
+          retry()
+        );
+        break;
+    }
+
+    this.subscriptions.add(process$.subscribe());
+
+    const command: Command<P, R> = new (class extends Command {
+      execute = (parameters: P) => {
```
Comment (znikola) [inline projects/core/src/util/command-query/command.service.ts:83]: https://sap-cx.slack.com/archives/DCTJGPDFV/p1616666589059200

File: projects/core/src/util/command-query/command.service.ts [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+        const result = new ReplaySubject<R>();
```
Comment (znikola) [inline projects/core/src/util/command-query/command.service.ts]: https://sap-cx.slack.com/archives/DCTJGPDFV/p1616667482061500

File: feature-libs/user/profile/components/close-account/components/close-account-modal/close-account-modal.component.ts [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+  }
+
+  dismissModal(reason?: any): void {
+    this.modalService.dismissActiveModal(reason);
+  }
+
+  closeAccount() {
+    this.isLoading$.next(true);
+
+    this.userProfile.close().subscribe({
```
Comment (znikola) [inline feature-libs/user/profile/components/close-account/components/close-account-modal/close-account-modal.component.ts:69]: https://sap-cx.slack.com/archives/DCTJGPDFV/p1616669399063400

File: feature-libs/user/profile/core/facade/facade-providers.ts [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+import { UserProfileService } from './user-profile.service';
+import { UserRegisterService } from './user-register.service';
+import {
+  UserEmailFacade,
+  UserPasswordFacade,
+  UserProfileFacade,
+  UserRegisterFacade,
+} from '@spartacus/user/profile/root';
+
+export const facadeProviders = [
```
Comment (znikola) [inline feature-libs/user/profile/core/facade/facade-providers.ts]: Why not type this array to `Provider[]`?
Same goes for other providers.ts files, e.g. feature-libs/user/account/core/facade/facade-providers.ts

**Review by znikola — CHANGES_REQUESTED** [review]

File: feature-libs/user/account/core/connectors/user-account.connector.ts (hunk 1 above)
Comment (dunqan) [reply feature-libs/user/account/core/connectors/user-account.connector.ts]: done, thanks

File: feature-libs/user/package.json (hunk 2 above)
Comment (dunqan) [reply feature-libs/user/package.json]: 👍

File: feature-libs/user/profile/core/facade/facade-providers.ts (hunk 6 above)
Comment (dunqan) [reply feature-libs/user/profile/core/facade/facade-providers.ts]: ok

**Review by znikola — DISMISSED** [review]

File: feature-libs/user/account/assets/translations/translations.ts [hunk 7]
```diff
@@ -0,0 +1,10 @@
+import { TranslationChunksConfig, TranslationResources } from '@spartacus/core';
+import { en } from './en/index';
+
+export const userTranslations: TranslationResources = {
+  en,
+};
+
+export const userTranslationChunksConfig: TranslationChunksConfig = {
+  userAccount: ['loginForm'],
```
Comment (marlass) [inline feature-libs/user/account/assets/translations/translations.ts]: Should you add here `forgottenPassword`?

File: scripts/changelog.ts [hunk 8]
```diff
@@ -74,6 +74,7 @@ export default async function run(
     '@spartacus/assets': 'projects/assets',
     '@spartacus/schematics': 'projects/schematics',
     '@spartacus/incubator': 'projects/incubator',
+    '@spartacus/user': 'projects/user',
```
Comment (marlass) [inline scripts/changelog.ts]:
```suggestion
    '@spartacus/user': 'feature-libs/user',
```

File: feature-libs/user/account/assets/translations/translations.ts (hunk 7 above)
Comment (marlass) [reply feature-libs/user/account/assets/translations/translations.ts]: Or remove it from the translations file

File: feature-libs/user/profile/root/user-profile-root.module.ts [hunk 9]
```diff
@@ -2,12 +2,13 @@ import { NgModule } from '@angular/core';
...
   provideDefaultConfigFactory,
   UserProfileFacadeTransitionalToken,
+  UserRegisterFacadeTransitionalToken,
 } from '@spartacus/core';
 import {
   USER_PROFILE_CORE_FEATURE,
   USER_PROFILE_FEATURE,
 } from './feature-name';
-import { UserProfileFacade } from './facade';
+import { UserProfileFacade, UserRegisterFacade } from './facade';
```
Comment (znikola) [inline feature-libs/user/profile/root/user-profile-root.module.ts]: It's better to import straight from the file, or from the barrel (index.ts) file

**Review by znikola — CHANGES_REQUESTED** [review]

File: feature-libs/user/profile/root/user-profile-root.module.ts (hunk 9 above)
Comment (dunqan) [reply feature-libs/user/profile/root/user-profile-root.module.ts]: 👍

File: feature-libs/user/account/components/login/login.component.html [hunk 10]
```diff
@@ -0,0 +1,12 @@
+<ng-container *ngIf="user$ | async as user; else login">
+  <div class="cx-login-greet">
+    {{ 'miniLogin.userGreeting' | cxTranslate: { name: user.name } }}
```
Comment (marlass) [inline feature-libs/user/account/components/login/login.component.html:3]: I don't see this namespace in translations in this entry point

File: feature-libs/user/account/components/ng-package.json [hunk 11]
```diff
@@ -0,0 +1,12 @@
+{
+  "$schema": "../../../../node_modules/ng-packagr/ng-package.schema.json",
+  "lib": {
+    "entryFile": "./public_api.ts",
+    "umdModuleIds": {
+      "@spartacus/core": "core",
+      "@spartacus/storefront": "storefront",
+      "@spartacus/user/core": "user",
+      "@ng-select/ng-select": "ng-select"
```
Comment (marlass) [inline feature-libs/user/account/components/ng-package.json]: Is it referenced in this lib?

File: feature-libs/user/account/ng-package.json [hunk 12]
```diff
@@ -0,0 +1,11 @@
+{
+  "$schema": "../../../node_modules/ng-packagr/ng-package.schema.json",
+  "lib": {
+    "entryFile": "./public_api.ts",
+    "umdModuleIds": {
+      "@spartacus/core": "core",
+      "@spartacus/storefront": "storefront",
+      "@ng-select/ng-select": "ng-select"
```
Comment (marlass) [inline feature-libs/user/account/ng-package.json]: Same. Needed?

File: feature-libs/user/account/occ/adapters/config/occ-user-account-endpoint.model.ts [hunk 13]
```diff
@@ -0,0 +1,12 @@
+import { OccEndpoint } from '@spartacus/core';
+
+declare module '@spartacus/core' {
+  interface OccEndpoints {
+    /**
+     * Get user account details
+     *
+     * @member {string}
```
Comment (marlass) [inline feature-libs/user/account/occ/adapters/config/occ-user-account-endpoint.model.ts]: Do you know what this tag does?

File: feature-libs/user/account/root/facade/user-account.facade.ts [hunk 14]
```diff
@@ -0,0 +1,22 @@
...
+import { User } from '../model/user.model';
+import { facadeFactory } from '@spartacus/core';
+import { USER_ACCOUNT_CORE_FEATURE } from '../feature-name';
+
+export function userAccountFacadeFactory() {
+  return facadeFactory({
+    facade: UserAccountFacade,
+    feature: USER_ACCOUNT_CORE_FEATURE,
+    methods: ['get'],
+    async: true,
```
Comment (marlass) [inline feature-libs/user/account/root/facade/user-account.facade.ts]: Why this one is `async`? I thought it's only needed when using store

File: angular.json [hunk 15]
```diff
@@ -805,6 +809,43 @@
...
       }
+    },
+    "user": {
+      "projectType": "library",
+      "root": "feature-libs/user",
+      "sourceRoot": "feature-libs/user",
+      "prefix": "cx",
+      "architect": {
+        "build": {
+          "builder": "@angular-devkit/build-ng-packagr:build",
```
Comment (marlass) [inline angular.json]: Use the builder for augmented types?

**Review by marlass — CHANGES_REQUESTED** [review]

File: feature-libs/user/profile/assets/translations/en/user.ts [hunk 16]
```diff
@@ -0,0 +1,92 @@
...
+    emailAddress: {
+      label: 'Email address',
+      placeholder: 'Enter email',
+    },
+    password: {
+      label: 'Password',
+      placeholder: 'Password',
+    },
+    wrongEmailFormat: 'This is not a valid email format.',
+  },
```
Comment (marlass) [inline feature-libs/user/profile/assets/translations/en/user.ts]: Isn't it part of account entry point?

File: feature-libs/user/profile/components/close-account/components/close-account-modal/close-account-modal.component.ts [hunk 17]
```diff
@@ -0,0 +1,80 @@
...
+  isLoggedIn$: Observable<boolean>;
+  isLoading$ = new BehaviorSubject(false);
+
+  constructor(
+    protected modalService: ModalService,
+    private authService: AuthService,
+    private globalMessageService: GlobalMessageService,
+    private routingService: RoutingService,
+    private translationService: TranslationService,
+    private userProfile: UserProfileFacade
```
Comment (marlass) [inline feature-libs/user/profile/components/close-account/components/close-account-modal/close-account-modal.component.ts]: Why private?

File: feature-libs/user/profile/components/close-account/components/close-account/close-account.component.ts [hunk 18]
```diff
@@ -0,0 +1,19 @@
...
+import { CloseAccountModalComponent } from '../close-account-modal/close-account-modal.component';
+
+@Component({
+  selector: 'cx-close-account',
+  templateUrl: './close-account.component.html',
+  changeDetection: ChangeDetectionStrategy.OnPush,
+})
+export class CloseAccountComponent {
+  modal: any;
+  constructor(private modalService: ModalService) {}
```
Comment (marlass) [inline feature-libs/user/profile/components/close-account/components/close-account/close-account.component.ts]: private ?

File: feature-libs/user/profile/components/forgot-password/forgot-password.component.ts [hunk 19]
```diff
@@ -0,0 +1,62 @@
...
+        [Validators.required, CustomFormValidators.emailValidator],
+      ],
+    });
+  }
+
+  requestForgotPasswordEmail() {
+    if (this.forgotPasswordForm.valid) {
+      this.userPassword
+        .requestForgotPasswordEmail(this.forgotPasswordForm.value.userEmail)
+        .pipe()
```
Comment (marlass) [inline feature-libs/user/profile/components/forgot-password/forgot-password.component.ts]: 👀

File: feature-libs/user/profile/components/register/register.component.ts [hunk 20]
```diff
@@ -0,0 +1,213 @@
...
+      ?.registerConsent;
+
+    if (requiredConsents && registerConsent) {
+      return requiredConsents.includes(registerConsent);
+    }
+
+    return false;
+  }
+
+  private onRegisterUserSuccess(success: boolean): void {
```
Comment (marlass) [inline feature-libs/user/profile/components/register/register.component.ts]: Do we want to keep this argument?

File: feature-libs/user/profile/components/reset-password/reset-password-form.component.ts [hunk 21]
```diff
@@ -0,0 +1,88 @@
...
+        'password',
+        'repassword'
+      ),
+    }
+  );
+
+  constructor(
+    private fb: FormBuilder,
+    private routingService: RoutingService,
+    private userPassword: UserPasswordFacade,
```
Comment (marlass) [inline feature-libs/user/profile/components/reset-password/reset-password-form.component.ts]: protected maybe?

File: feature-libs/user/profile/components/reset-password/reset-password.module.ts [hunk 22]
```diff
@@ -0,0 +1,35 @@
...
+    ReactiveFormsModule,
+    RouterModule,
+    I18nModule,
+    FormErrorsModule,
+  ],
+  providers: [
+    provideDefaultConfig(<CmsConfig>{
+      cmsComponents: {
+        ResetPasswordComponent: {
+          component: ResetPasswordFormComponent,
```
Comment (marlass) [inline feature-libs/user/profile/components/reset-password/reset-password.module.ts:27]: What do you think about renaming component  and drop the `form` from it's name?

File: feature-libs/user/profile/components/update-email/update-email.component.ts [hunk 23]
```diff
@@ -0,0 +1,57 @@
...
+@Component({
+  selector: 'cx-update-email',
+  templateUrl: './update-email.component.html',
+})
+export class UpdateEmailComponent {
+  constructor(
+    private routingService: RoutingService,
+    private globalMessageService: GlobalMessageService,
+    private userEmail: UserEmailFacade,
+    private authService: AuthService
```
Comment (marlass) [inline feature-libs/user/profile/components/update-email/update-email.component.ts]: protected ?

File: feature-libs/user/profile/components/update-password/components/update-password/update-password.component.ts [hunk 24]
```diff
@@ -0,0 +1,53 @@
...
+    newPassword,
+  }: {
+    oldPassword: string;
+    newPassword: string;
+  }): void {
+    this.isLoading$.next(true);
+    this.userPassword.update(oldPassword, newPassword).subscribe({
+      next: () => this.onSuccess(),
+      error: () => {
+        this.isLoading$.next(false);
```
Comment (marlass) [inline feature-libs/user/profile/components/update-password/components/update-password/update-password.component.ts:46]: Looks not needed

File: feature-libs/user/profile/components/update-password/components/update-password/update-password.component.ts [hunk 25]
```diff
@@ -0,0 +1,53 @@
...
+  selector: 'cx-update-password',
+  templateUrl: './update-password.component.html',
+})
+export class UpdatePasswordComponent {
+  isLoading$ = new BehaviorSubject(false);
+
+  constructor(
+    private routingService: RoutingService,
+    private userPassword: UserPasswordFacade,
+    private globalMessageService: GlobalMessageService
```
Comment (marlass) [inline feature-libs/user/profile/components/update-password/components/update-password/update-password.component.ts]: protected ?

**Review by marlass — CHANGES_REQUESTED** [review]

File: feature-libs/user/profile/components/update-password/components/update-password/update-password.component.ts (hunk 25 above)
Comment (dunqan) [reply feature-libs/user/profile/components/update-password/components/update-password/update-password.component.ts]: 👍

File: feature-libs/user/profile/components/update-password/components/update-password/update-password.component.ts (hunk 24 above)
Comment (dunqan) [reply feature-libs/user/profile/components/update-password/components/update-password/update-password.component.ts:46]: Actually, it is, as on error complete is not called.

File: feature-libs/user/profile/components/update-email/update-email.component.ts (hunk 23 above)
Comment (dunqan) [reply feature-libs/user/profile/components/update-email/update-email.component.ts]: 👍

File: feature-libs/user/profile/components/reset-password/reset-password.module.ts (hunk 22 above)
Comment (dunqan) [reply feature-libs/user/profile/components/reset-password/reset-password.module.ts:27]: This component is undergoing follow up refactor in #10443 / #11660

File: feature-libs/user/profile/components/reset-password/reset-password-form.component.ts (hunk 21 above)
Comment (dunqan) [reply feature-libs/user/profile/components/reset-password/reset-password-form.component.ts]: 👍

File: feature-libs/user/profile/components/register/register.component.ts (hunk 20 above)
Comment (dunqan) [reply feature-libs/user/profile/components/register/register.component.ts]: Nope 👍

File: feature-libs/user/profile/components/forgot-password/forgot-password.component.ts (hunk 19 above)
Comment (dunqan) [reply feature-libs/user/profile/components/forgot-password/forgot-password.component.ts]: :-) 👍

File: feature-libs/user/profile/components/close-account/components/close-account/close-account.component.ts (hunk 18 above)
Comment (dunqan) [reply feature-libs/user/profile/components/close-account/components/close-account/close-account.component.ts]: 👍

File: feature-libs/user/profile/components/close-account/components/close-account-modal/close-account-modal.component.ts (hunk 17 above)
Comment (dunqan) [reply feature-libs/user/profile/components/close-account/components/close-account-modal/close-account-modal.component.ts]: 👍 Fixed.

File: feature-libs/user/profile/assets/translations/en/user.ts (hunk 16 above)
Comment (dunqan) [reply feature-libs/user/profile/assets/translations/en/user.ts]: Nice catch! It's a duplicate. Thanks!

File: feature-libs/user/account/root/facade/user-account.facade.ts (hunk 14 above)
Comment (dunqan) [reply feature-libs/user/account/root/facade/user-account.facade.ts]: Correct, removed.

File: feature-libs/user/account/occ/adapters/config/occ-user-account-endpoint.model.ts (hunk 13 above)
Comment (dunqan) [reply feature-libs/user/account/occ/adapters/config/occ-user-account-endpoint.model.ts]: https://jsdoc.app/tags-member.html :-) 
Removed 👍

File: feature-libs/user/account/components/ng-package.json (hunk 11 above)
Comment (dunqan) [reply feature-libs/user/account/components/ng-package.json]: Thanks, removed all occurrences.

File: angular.json (hunk 15 above)
Comment (dunqan) [reply angular.json]: 👍

File: feature-libs/user/account/ng-package.json (hunk 12 above)
Comment (dunqan) [reply feature-libs/user/account/ng-package.json]: Removed.

File: scripts/changelog.ts (hunk 8 above)
Comment (dunqan) [reply scripts/changelog.ts]: 🤦 👍

File: feature-libs/user/account/assets/translations/translations.ts (hunk 7 above)
Comment (dunqan) [reply feature-libs/user/account/assets/translations/translations.ts]: Removed. Thanks!

File: feature-libs/user/account/components/login/login.component.html (hunk 10 above)
Comment (dunqan) [reply feature-libs/user/account/components/login/login.component.html:3]: Moved.

File: feature-libs/user/profile/assets/translations/translations.ts [hunk 26]
```diff
@@ -0,0 +1,10 @@
+import { TranslationChunksConfig, TranslationResources } from '@spartacus/core';
+import { en } from './en/index';
+
+export const userProfileTranslations: TranslationResources = {
+  en,
+};
+
+export const userTranslationChunksConfig: TranslationChunksConfig = {
+  user: ['updateEmailForm', 'register', 'loginForm', 'forgottenPassword'],
```
Comment (marlass) [inline feature-libs/user/profile/assets/translations/translations.ts]:
```suggestion
  user: ['updateEmailForm', 'register', 'forgottenPassword'],
```

I think this should not be here

File: feature-libs/user/profile/components/update-profile/update-profile.component.ts [hunk 27]
```diff
@@ -0,0 +1,47 @@
...
+})
+export class UpdateProfileComponent {
+  titles$: Observable<Title[]> = this.userProfile.getTitles();
+  user$: Observable<User | undefined> = this.userProfile.get();
+  isLoading$ = new BehaviorSubject(false);
+
+  constructor(
+    private routingService: RoutingService,
+    private userProfile: UserProfileFacade,
+    private globalMessageService: GlobalMessageService
```
Comment (marlass) [inline feature-libs/user/profile/components/update-profile/update-profile.component.ts]: protected?

File: feature-libs/user/profile/components/update-profile/update-profile.component.ts [hunk 28]
```diff
@@ -0,0 +1,47 @@
...
+  onCancel(): void {
+    this.routingService.go({ cxRoute: 'home' });
+  }
+
+  onSubmit({ userUpdates }: { userUpdates: User }): void {
+    this.isLoading$.next(true);
+
+    this.userProfile.update(userUpdates).subscribe({
+      next: () => this.onSuccess(),
+      error: () => {},
```
Comment (marlass) [inline feature-libs/user/profile/components/update-profile/update-profile.component.ts:43]: In this case shouldn't we also disable loading on error?

File: projects/core/src/occ/adapters/user/occ-user.adapter.ts [hunk 29]
```diff
@@ -17,6 +17,10 @@ import {
   InterceptorUtil,
   USE_CLIENT_TOKEN,
 } from '../../utils/interceptor-util';
+/**
+ * @deprecated since 3.2, use `OccUserDetailAdapter` from the `@spartacus/user/detail`
+ * package or `OccUserProfileAdapter` from  the `@spartacus/user/detail` package.
```
Comment (marlass) [inline projects/core/src/occ/adapters/user/occ-user.adapter.ts]: Packages to be renamed here to Profile or Account?

File: projects/core/src/user/connectors/user/user.adapter.ts [hunk 30]
```diff
@@ -1,6 +1,9 @@
 import { Observable } from 'rxjs';
 import { Title, User, UserSignUp } from '../../../model/misc.model';
-
+/**
+ * @deprecated since 3.2, use `OccUserDetailAdapter` from the `@spartacus/user/detail`
+ * package or `OccUserProfileAdapter` from  the `@spartacus/user/detail` package.
```
Comment (marlass) [inline projects/core/src/user/connectors/user/user.adapter.ts]: package name to updates

File: projects/core/src/user/store/effects/user-details.effect.ts [hunk 31]
```diff
@@ -13,6 +13,7 @@ export class UserDetailsEffects {
   loadUserDetails$: Observable<UserActions.UserDetailsAction> = this.actions$.pipe(
     ofType(UserActions.LOAD_USER_DETAILS),
     map((action: UserActions.LoadUserDetails) => action.payload),
+    tap(() => console.log('old load user effect')),
```
Comment (marlass) [inline projects/core/src/user/store/effects/user-details.effect.ts]:
```suggestion
```

File: projects/core/src/user/store/reducers/transitoinal.ts [hunk 32]
```diff
@@ -0,0 +1,120 @@
+import { Provider } from '@angular/core';
```
Comment (marlass) [inline projects/core/src/user/store/reducers/transitoinal.ts]: Typo in filename

File: projects/storefrontapp/src/app/app.module.ts [hunk 33]
```diff
@@ -100,6 +100,12 @@ if (!environment.production) {
     ...devImports,
   ],
 
+  providers: [
+    provideConfig({
+      icon: { symbols: { MIC: 'fas fa-microphone' } },
+    } as IconConfig),
+  ],
```
Comment (marlass) [inline projects/storefrontapp/src/app/app.module.ts]: Where it's used?

**Review by marlass — CHANGES_REQUESTED** [review]

File: feature-libs/user/profile/assets/translations/translations.ts (hunk 26 above)
Comment (dunqan) [reply feature-libs/user/profile/assets/translations/translations.ts]: 👍

File: feature-libs/user/profile/components/update-profile/update-profile.component.ts (hunk 27 above)
Comment (dunqan) [reply feature-libs/user/profile/components/update-profile/update-profile.component.ts]: 👍

File: projects/core/src/user/store/effects/user-details.effect.ts (hunk 31 above)
Comment (dunqan) [reply projects/core/src/user/store/effects/user-details.effect.ts]: 😊

File: projects/core/src/user/store/reducers/transitoinal.ts (hunk 32 above)
Comment (dunqan) [reply projects/core/src/user/store/reducers/transitoinal.ts]: 👍

File: feature-libs/user/profile/components/update-profile/update-profile.component.ts (hunk 28 above)
Comment (dunqan) [reply feature-libs/user/profile/components/update-profile/update-profile.component.ts:43]: We don't have an onError scenario here, will be fixed in the followup #11662

File: projects/core/src/occ/adapters/user/occ-user.adapter.ts (hunk 29 above)
Comment (dunqan) [reply projects/core/src/occ/adapters/user/occ-user.adapter.ts]: To both actually, thanks!

File: projects/core/src/user/connectors/user/user.adapter.ts (hunk 30 above)
Comment (dunqan) [reply projects/core/src/user/connectors/user/user.adapter.ts]: Fixed

File: projects/storefrontapp/src/app/app.module.ts (hunk 33 above)
Comment (dunqan) [reply projects/storefrontapp/src/app/app.module.ts]: Dunno 🤷 
Thanks ;-)

**Review by marlass — DISMISSED** [review]

**Review by znikola — DISMISSED** [review]

**Review by marlass — APPROVED** [review]

**Review by znikola — APPROVED** [review]

**Pull request merged**